### PR TITLE
 Introducing noitatsxoB

### PR DESCRIPTION
**Pull request by silicons**
## About The Pull Request

hahaha sike I'm not a mapper (yet)
image

No but adds map rotation in the literal sense of the word. I'll probably PR random mining ruin rotation in another PR.

## Why It's Good For The Game

More possibilities, etc etc. Disabled by default ofcourse, I'm not that cruel ;)

## Changelog
:cl:
add: The map loader is now rotational.
/:cl:

### CODE REVIEW

**Comment by silicons** [comment]
OH SHIT IT WORKS

**Comment by Linzolle** [comment]
this is cursed

**Comment by Putnam3145** [comment]
what the fuck is a Boxbatstoin

**Comment by MrJWhit** [comment]
Hey, this is _awful_ to play in. I can understand it being used in something like a wizard or an adminbus round where it's meant to be confusing and disorientating. But a regular-ass round? No, please stop it.

There's a minor issues with decals not rotating while in the center of the map making weird visual glitches.

But mostly, this is pain and I hate it.

**Comment by lolman360** [comment]
ah yes, Boxbatstoin, my favorite station

**Comment by silicons** [comment]
@MrJWhit support will be left in but this isn't for rotating primary maps

the ongoing stuff is mostly just testing.

in the future the only plans to actually use this will be for lavaland and similar ruins, and I guess when an admin feels like being funny.

File: code/modules/mapping/map_template.dm [hunk 1]
```diff
@@ -55,12 +87,12 @@
 	SSmachines.setup_template_powernets(cables)
 	SSair.setup_template_machinery(atmos_machines)
 
-/datum/map_template/proc/load_new_z(list/traits = list(ZTRAIT_AWAY = TRUE))
-	var/x = round((world.maxx - width)/2)
-	var/y = round((world.maxy - height)/2)
+/datum/map_template/proc/load_new_z(orientation = SOUTH, list/ztraits = src.ztraits || list(ZTRAIT_AWAY = TRUE), centered = TRUE)
+	var/x = centered? max(FLOOR((world.maxx - width) / 2, 1), 1) : 1
+	var/y = centered? max(FLOOR((world.maxy - height) / 2, 1), 1) : 1
```
Comment (Ghommie) [inline code/modules/mapping/map_template.dm]:
```suggestion
	var/x = centered? max(round((world.maxx - width) / 2), 1) : 1
	var/y = centered? max(round((world.maxy - height) / 2), 1) : 1
```
round(value) without second arg already acts like FLOOR(value, 1)

File: code/modules/mapping/map_template.dm [hunk 2]
```diff
@@ -108,19 +176,36 @@
...
+			deleted_atoms++
+	log_world("Annihilated [deleted_atoms] objects.")
 
 //for your ever biggening badminnery kevinz000
 //❤ - Cyberboss
-/proc/load_new_z_level(file, name, list/traits)
+/proc/load_new_z_level(file, name, orientation, list/ztraits)
 	var/datum/map_template/template = new(file, name)
-	return template.load_new_z(traits)
+	template.load_new_z(orientation, ztraits)
```
Comment (Ghommie) [inline code/modules/mapping/map_template.dm]:
```suggestion
	return template.load_new_z(orientation, ztraits)
```
The return value is used in createRandomZlevel to announce wheter the zlevel was successfully loaded or not.

File: code/modules/mapping/map_template.dm [hunk 3]
```diff
@@ -108,19 +176,36 @@
...
+
+/datum/map_template/proc/get_affected_turfs(turf/T, centered = FALSE, orientation = SOUTH)
+	var/turf/placement = T
+	if(centered)
+		var/turf/corner = locate(placement.x - FLOOR(((orientation & (NORTH|SOUTH))? width : height) / 2, 1), placement.y - FLOOR(((orientation & (NORTH|SOUTH))? height : width) / 2, 1), placement.z) // %180 catches East/West (90,270) rotations on true, North/South (0,180) rotations on false
```
Comment (Ghommie) [inline code/modules/mapping/map_template.dm]:
```suggestion
		var/turf/corner = locate(placement.x - round(((orientation & (NORTH|SOUTH))? width : height) / 2), placement.y - round(((orientation & (NORTH|SOUTH))? height : width) / 2), placement.z) // %180 catches East/West (90,270) rotations on true, North/South (0,180) rotations on false
```

File: code/modules/mapping/map_template.dm [hunk 4]
```diff
@@ -71,31 +103,67 @@
...
+  * * annihilate - Should we destroy stuff in our bounds while loading
+  * * force_cache - Should we force the parsed shuttle to cache instead of being GC'd post loading if it wasn't going to be cached by default
+  * * rotate_placement_to_orientation - Has no effect if centered. Should we rotate where we load it around the turf we're loading at? Used for stuff like engine submaps when the station is rotated.
+  *
+  */
+/datum/map_template/proc/load(turf/T, centered = FALSE, orientation = SOUTH, annihilate = default_annihilate, force_cache = FALSE, rotate_placement_to_orientation = FALSE)
+	var/old_T = T
 	if(centered)
-		T = locate(T.x - round(width/2) , T.y - round(height/2) , T.z)
+		T = locate(T.x - FLOOR(((orientation & (NORTH|SOUTH))? width : height) / 2, 1) , T.y - FLOOR(((orientation & (NORTH|SOUTH)) ? height : width) / 2, 1) , T.z) // %180 catches East/West (90,270) rotations on true, North/South (0,180) rotations on false
```
Comment (Ghommie) [inline code/modules/mapping/map_template.dm]:
```suggestion
		T = locate(T.x - round(((orientation & (NORTH|SOUTH))? width : height) / 2) , T.y - round(((orientation & (NORTH|SOUTH)) ? height : width) / 2) , T.z) // %180 catches East/West (90,270) rotations on true, North/South (0,180) rotations on false
```
Feel free to disregard these suggestions if the FLOOR macro turns out to be more efficient than the integrated round() function.

**Review by Ghommie — CHANGES_REQUESTED** [review]

**Review by Ghommie — APPROVED** [review]

**Pull request merged**